### PR TITLE
(fix) O3-2654: Unable to edit visit attributes when editing visit details

### PR DESCRIPTION
**Pull request by usamaidrsk**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

- Fix error on updating visit attributes, by adding separate handler for visit attribute update.

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/53287480/94917381-633c-4b1c-b07f-62686cf323ab

## Related Issue

[O3-2654](https://openmrs.atlassian.net/browse/O3-2654)

### CODE REVIEW

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 1]
```diff
@@ -316,6 +317,38 @@ const StartVisitForm: React.FC<StartVisitFormProps> = ({
...
 
       if (isOnline) {
+        if (visitToEdit?.uuid) {
+          const existingAttributes = visitToEdit.attributes.map((attr) => attr.attributeType.uuid) || [];
+          const updatedAttributes = [];
+          const addedAttributes = [];
+          payload.attributes.forEach((attribute) => {
+            if (existingAttributes.includes(attribute.attributeType)) {
+              const attributeUuid = visitToEdit.attributes.filter(
+                (attr) => attr.attributeType.uuid === attribute.attributeType,
```
Comment (vasharma05) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:327]: Fix: `attribute.attributeType.uuid`

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 2]
```diff
@@ -316,6 +317,38 @@ const StartVisitForm: React.FC<StartVisitFormProps> = ({
...
+                value: attribute.value,
+              },
+            }).catch((err) => {
+              showSnackbar({
+                title: t('errorUpdatingVisitAttribute', 'Error updating visit attribute'),
+                kind: 'error',
+                isLowContrast: false,
+                subtitle: err?.message,
+              });
+            });
```
Comment (vasharma05) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:349]: Why are we updating the visit details even before we are updating the visit details. 

The right solution will be to first update visit details and then if the request is successful, then we update the attributes.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 3]
```diff
@@ -316,6 +317,38 @@ const StartVisitForm: React.FC<StartVisitFormProps> = ({
...
+          const existingAttributes = visitToEdit.attributes.map((attr) => attr.attributeType.uuid) || [];
+          const updatedAttributes = [];
+          const addedAttributes = [];
+          payload.attributes.forEach((attribute) => {
+            if (existingAttributes.includes(attribute.attributeType)) {
+              const attributeUuid = visitToEdit.attributes.filter(
+                (attr) => attr.attributeType.uuid === attribute.attributeType,
+              )[0].uuid;
+              updatedAttributes.push({ uuid: attributeUuid, value: attribute.value });
+            } else addedAttributes.push(attribute);
```
Comment (vasharma05) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:330]: There are so many things that can fail here:
1. If attributes is undefined
2. If filter returns empty
3. If the request to update the visit attribute fails, this is assuming a very happy case, which is not always the case.

**Review by vasharma05 — CHANGES_REQUESTED** [review]

**Comment by vasharma05** [comment]
Hi @dkayiwa !
Do you know why are we not able to update the visit attributes in the visit update payload itself? Is there anything we can work out on this? Updating the visit attributes separately will bring up many edge cases.
For more info, please feel free to refer the ticket assigned.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx (hunk 1 above)
Comment (usamaidrsk) [reply packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:327]: `attribute` is already used in the scope @vasharma05

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx (hunk 3 above)
Comment (usamaidrsk) [reply packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:330]: > There are so many things that can fail here:
> 
> 1. If attributes is undefined
> 2. If filter returns empty
> 3. If the request to update the visit attribute fails, this is assuming a very happy case, which is not always the case.

- Upon failure, I added have a snackbar error feedback.

**Comment by ibacher** [comment]
@vasharma05 @usamaidrsk The way to deal with issues like this would be to open in a Jira ticket in the Webservices.REST project that says something like: "Add ability to update attributes as part of the visit payload" with a description of what you're trying to accomplish and what syntax should be supported. You can add the tag "community-priority" in Jira and it should show up on the appropriate board. In most cases like these you're simply looking for a feature no one has ever asked for before, so ask for the feature (which is what a ticket is for).

**Comment by vasharma05** [comment]
Hi @ibacher!
I actually thought that this might be a part of the well thought design, rather than not being implemented since no one asked for it. Bad assumption from my side. 

I'll create a ticket and update the ticket in the O3-2654.
Thanks a lot @ibacher !

**Comment by vasharma05** [comment]
Expecting a better solution with REST Ticket: [https://openmrs.atlassian.net/browse/RESTWS-926](https://openmrs.atlassian.net/jira/software/c/projects/RESTWS/boards/59?selectedIssue=RESTWS-926)

**Comment by ibacher** [comment]
@vasharma05 hmmm… actually there is an issue with updates vs creates and it comes down to ambiguity in JSON. Ideally, what we want to submit as an edit is something that records the changes we want to make. There are some specified syntaxes for this (JSON Patch and XML Patch) which we don’t support. Without either an explicit patching syntax or sending the full resource with each request, the backend has no way to differentiate between an “append this attribute” request vs a “delete all attributes but this one” request or even a “delete all attributes”.

Theoretically, the principles of REST mean that we should be sending the entire representation of the resource as part of the PUT request, so we do need to be sure (on the front end side) that the update includes the full visit payload with the modifications made.

**Comment by vasharma05** [comment]
This is what I also encountered in the patient registration, when dealing with identifiers. If you see, there were cases handled where a few identifiers were updated, deleted and added simultaneously, which I belive cannot be done in a single REST request, hence I asked @usamaidrsk to follow the same steps here as well.

**Pull request closed without merging**